google/cloud/aiplatform/docker_utils/build.py

```diff
@@ -0,0 +1,430 @@
+        ret += "EXPOSE {}".format(port)
+    return ret
+
+
+def make_dockerfile(
+    base_image: str,
+    main_package: Package,
+    container_workdir: str,
+    container_home: str,
+    requirements_path: Optional[str] = None,
+    setup_path: Optional[str] = None,
+    extra_requirements: Optional[List[str]] = None,
+    extra_packages: Optional[List[str]] = None,
+    extra_dirs: Optional[List[str]] = None,
+    exposed_ports: Optional[List[int]] = None,
+    pip_command: str = "pip",
+    python_command: str = "python",
+) -> str:
+    """Generates a Dockerfile for building an image.
+
+    It builds on a specified base image to create a container that:
+    - installs any dependency specified in a requirements.txt or a setup.py file,
+    and any specified dependency packages existing locally or found from PyPI
+    - copies all source needed by the main module, and potentially injects an
+    entrypoint that, on run, will run that main module
+
+    Args:
+        base_image (str):
+            Required. The ID or name of the base image to initialize the build stage.
+        main_package (Package):
+            Required. The main application to execute.
+        container_workdir (str):
+            Required. The working directory in the container.
+        container_home (str):
+            Required. The $HOME directory in the container.
+        requirements_path (str):
+            Optional. The path to a local requirements.txt file.
+        setup_path (str):
+            Optional. The path to a local setup.py file.
+        extra_requirements (List[str]):
+            Optional. The list of required dependencies to install from PyPI.
+        extra_packages (List[str]):
+            Optional. The list of user custom dependency packages to install.
+        extra_dirs: (List[str]):
+            Optional. The directories other than the work_dir required to be in the container.
+        exposed_ports (List[int]):
+            Optional. The exposed ports that the container listens on at runtime.
+        pip_command (str):
+            Required. The pip command used for install packages.
+        python_command (str):
+            Required. The python command used for running python code.
+
+    Returns:
+        A string that represents the content of a Dockerfile.
+    """
+    dockerfile = textwrap.dedent(
+        """
+        FROM {base_image}
+
+        # The directory is created by root. This sets permissions so that any user can
+        # access the folder.
+        RUN mkdir -m 777 -p {workdir} {container_home}
+        WORKDIR {workdir}
+        ENV HOME={container_home}
+
+        # Keeps Python from generating .pyc files in the container
+        ENV PYTHONDONTWRITEBYTECODE=1
+        """.format(
+            base_image=base_image,
+            workdir=shlex_quote(container_workdir),
+            container_home=shlex_quote(container_home),
+        )
+    )
+
+    dockerfile += _prepare_dependency_entries(
```

Comment:
They need to install `google-cloud-aiplatform[prediction]`.

In the experimental launch, since we are using custom-prediction-routine branch, they need to specify
`google-cloud-aiplatform[prediction] @ git+https://github.com/googleapis/python-aiplatform.git@custom-prediction-routine`

Reply:
Do we expect the users to include `google-cloud-aiplatform[prediction]` in their requirements file? What if it's missing?

Reply:
I expect that :). We could consider adding a default to be the latest version of SDK.
In experimental launch, we use a different branch so it may decrease the opportunity that users want to install a specific version of SDK. However, after merging to main branch, users likely need to choose one specific version that works for them. WDYT?

Reply:
I'm in favor of installing the custom-prediction-routine branch of the SDK on behalf of the users for the experimental, since I suspect many users will forget to add this to their dockerfile.

We can decide later for GA if we want to force upgrade the user's version if their version is too outdate, or if we want the user to resolve sdk versions themselves.

Reply:
Updated.